src/main/java/org/template/recommendation/Algorithm.java

```diff
@@ -128,11 +140,218 @@ public FilterCorrelators(String actionName, List<String> itemIDs) {
+
+            cooccurrenceIDS = SimilarityAnalysisJava.crossOccurrenceDownsampled(
+                    datasets,
+                    ap.getSeed() == null ?  (int)System.currentTimeMillis(): ap.getSeed().intValue());
+
+        }
+
+        List<Tuple2<String, IndexedDataset>> cooccurrenceCorrelators =
+                new ArrayList<>();
+
+        for(int i=0; i<cooccurrenceIDS.size(); i++){
+            cooccurrenceCorrelators.add(new Tuple2<>(
+                    preparedData.getActions().get(i)._1(),
+                    cooccurrenceIDS.get(i)
+            ));
+        }
+
+        JavaPairRDD<String, Map<String, JsonAST.JValue>> propertiesRDD;
+        if (calcPopular) {
+            JavaPairRDD<String, Map<String, JsonAST.JValue>> ranksRdd = getRanksRDD(preparedData.getFieldsRDD(), sc);
+            propertiesRDD = preparedData.getFieldsRDD().fullOuterJoin(ranksRdd).mapToPair(new CalcAllFunction());
+        } else {
+            propertiesRDD = RDDUtils.getEmptyPairRDD(sc);
+        }
+
+        logger.info("Correlators created now putting into URModel");
+
+        // singleton list for propertiesRdd
+        ArrayList<JavaPairRDD<String, Map<String, JsonAST.JValue>>> pList = new ArrayList<>();
+        pList.add(propertiesRDD);
+        new URModel(
+//                cooccurrenceCorrelators,
+                null,
+                pList,
+                getRankingMapping(),
+                false,
+                sc).save(dateNames, esIndex, esType);
+        return new NullModel();
+    }
+
+    private Map<String,String> getRankingMapping(){
+        HashMap<String, String> out = new HashMap<>();
+        for (String r: rankingFieldNames){
+            out.put(r, "float");
+        }
+        return out;
+    }
+
+    /**
+     * Lambda expression class for calcPopular in calcAll()
+     */
+    private static class CalcAllFunction implements
+            PairFunction<
+                    Tuple2<String,Tuple2<Optional<Map<String,JsonAST.JValue>>,Optional<Map<String, JsonAST.JValue>>>>,
+                    String,
+                    Map<String,JsonAST.JValue>
+                    >{
+        public Tuple2<String,Map<String, JsonAST.JValue>> call(
+                Tuple2<String,Tuple2<Optional<Map<String,JsonAST.JValue>>,Optional<Map<String, JsonAST.JValue>>>> t){
 
+            String item = t._1();
+            Optional<Map<String, JsonAST.JValue>> oFieldsPropMap = t._2()._1();
+            Optional<Map<String, JsonAST.JValue>> oRankPropMap = t._2()._2();
+
+            if (oFieldsPropMap.isPresent() && oRankPropMap.isPresent()){
+                Map<String, JsonAST.JValue> fieldPropMap = oFieldsPropMap.get();
+                Map<String, JsonAST.JValue> rankPropMap = oRankPropMap.get();
+                HashMap<String, JsonAST.JValue> newMap = new HashMap<>(fieldPropMap);
+                newMap.putAll(rankPropMap);
+                return new Tuple2<>(item, newMap);
+
+            }else if (oFieldsPropMap.isPresent()){
```

Comment:
@jk2227 direct to @zhuyifan1996 in his pull req for everything that is not conversions.java